pilot/pkg/xds/fake.go

```diff
@@ -249,12 +243,21 @@ func NewFakeDiscoveryServer(t test.Failer, opts FakeOptions) *FakeDiscoveryServe
 	s.Start(stop)
 
 	se.ResyncEDS()
-	if err := k8s.ResyncEndpoints(); err != nil {
+	if err := k8s.ForceResync(); err != nil {
+		t.Fatal(err)
+	}
+
+	s.updateMutex.Lock()
+	defer s.updateMutex.Unlock()
+	ctx := model.NewPushContext()
+	if err := ctx.InitContext(env, env.PushContext, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err := s.UpdateServiceShards(ctx); err != nil {
 		t.Fatal(err)
 	}
+	env.PushContext = ctx
```

Comment:
Do we really need this?

Reply:
Yes. Without this we get the flakes. If I just call InitContext on the existing one, there are races.